### PR TITLE
Upgrade Phing to version 3.0.0.

### PR DESCRIPTION
**Pull request by EreMaijala**
This upgrades Phing to version 3.0.0 and modifies the build file accordingly. I've tried to test as much as I can, but there are still several tasks that I cannot test (e.g. Selenium).

### CODE REVIEW

File: composer.json
Comment (demiankatz) [inline composer.json]: Did you mean to roll this back?

File: build.xml
Comment (demiankatz) [inline build.xml]: Should we drop the zend.enable_gc comment entirely? I suspect it no longer applies...

File: build.xml
Comment (demiankatz) [inline build.xml:372]: Feels like it might be beneficial at some point to refactor so we don't have to copy and paste all these variables to so many different places, but that's a project for another day... just thinking out loud!

File: build.xml
Comment (demiankatz) [inline build.xml:749]: Would be nice if they had an equivalent to the `-p` switch!

File: build.xml
Comment (demiankatz) [inline build.xml]: Any idea why this commented-out line is here? Maybe now is a chance to clean up old junk.

**Review by demiankatz — CHANGES_REQUESTED** [review]
Thanks, @EreMaijala, this looks pretty reasonable to me; see below for a few comments.

I haven't actually tried it yet, but I can do so after you make any needed revisions.

I'm not too worried about Selenium -- I doubt it works anymore anyway, and I haven't used it in years. But if there's anything else you'd like me to help test specifically, please let me know!

File: build.xml
Comment (EreMaijala) [reply build.xml:372]: I agree, but it also complicates things. I wish there was support for a set of env vars...

File: build.xml
Comment (EreMaijala) [reply build.xml:749]: Agreed. It already works like `-p`, but spits out scary-looking error messages (it did show warnings before as well, but not so scary-looking).

File: build.xml
Comment (EreMaijala) [reply build.xml]: Yes, dropped!

File: composer.json
Comment (EreMaijala) [reply composer.json]: Certainly not. No idea how I managed to do it without conflicts..

File: build.xml
Comment (EreMaijala) [reply build.xml]: No idea. Now removed.

**Comment by EreMaijala** [comment]
@demiankatz I've now tested most targets, but `package` is the one that could use more testing.

**Review by demiankatz — CHANGES_REQUESTED** [review]
Thanks, @EreMaijala. I tested the `package` task and found that it was broken due to a single missing dollar sign. I've pushed up a fix, and also done a regex search to be sure the error is not repeated anywhere else in the file (it isn't).

Running the task results in comparable results to what I get from Phing2 (though the file sizes are slightly different because Phing itself is part of the bundle, and the new version is a different size than the old version).

I think this is probably safe to merge, but I'm going to hold off on hitting the merge button in case there's anything else you'd like to try first. But if you're comfortable, please go ahead and merge at your convenience, and we'll see if Jenkins turns up any new problems...

(And then there's all the subprojects to deal with... *sigh*).

**Review by demiankatz — APPROVED** [review]
Sorry, meant to approve, not request changes. Force of habit!

**Review by demiankatz — CHANGES_REQUESTED** [review]
Hmm, maybe we're not ready yet. When I finished up work for the morning, I ran the `shutdown` task and got this:

```
vufind > shutdown:

   [delete] Could not find file /etc/apache2/conf-enabled/vufindtest.conf to delete.
     [exec] Result: 127
     [exec] Result: 1
     [exec] Result: 1
No process found for Solr node running on port 8983
     [exec] Result: 1
     [exec] Result: 127
   [delete] Deleting directory /usr/local/vufind/solr
   [delete] Deleting directory /usr/local/vufind/vendor
   [delete] Deleting directory /usr/local/vufind/local

BUILD FAILED
Class "Phing\Io\IOException" not found
```

Are you seeing the same thing?

Maybe the problem occurs if you run it twice in a row, or if some parts of it aren't necessary -- e.g. shutting down after a failed partial startup. It's possible I double shut-down by accident. But in any case, it left things in a scrambled state, which is not what usually happens.

**Comment by EreMaijala** [comment]
@demiankatz I can reproduce by running shutdown without startup first, apart from the last `Class "Phing\Io\IOException" not found`, which I assume would be the result of the vendor directory going missing from under it. I believe the behavior is the same as before, it's just that Phing now emits the scary-looking errors about exec statuses (and looking at the source there's no way to suppress them).

I've pushed a set of tweaks and checks to alleviate the issue.

**Comment by demiankatz** [comment]
Thanks, @EreMaijala. I think your theory about why I'm getting the missing class message makes sense, though it's strange that you're not getting the same thing.

Unfortunately, even with these changes, I'm still getting left in a bad state when I shutdown at the wrong time, which then breaks my ability to start up again (I think it bails out before the git reset, so then a bunch of files are missing).

Do you think perhaps it is time to stop deleting the vendor directory? I believe that we've talked about this in the past -- it's not necessarily helpful anyway, as long as startup runs composer to update things properly. It would save a lot of time if we didn't have to keep uninstalling and reinstalling all the dependencies... and if it's now breaking Phing, that's an excuse to make the change.

On a side note, I also notice that you pinned Composer at 2.7.1; is there a particular reason for that, given that newer versions are available?

**Comment by EreMaijala** [comment]
@demiankatz Composer version was a mistake, now pinned to latest. I've removed deletion of vendor and added `composer install` to startup. Does this work any better for you?

File: build.xml
Comment (demiankatz) [inline build.xml]: Isn't the composer target already calling composer install? Do we need to explicitly call it again here?

I also wonder if we should be downloading composer.phar into ${srcdir}/downloads so it's not cluttering up the root of the project... though that's probably just scope creep at this point and is better addressed in a follow-up. :-)

**Review by demiankatz — CHANGES_REQUESTED** [review]
Thanks, @EreMaijala -- this seems to be working better for me, but see below for one more question.

I think whatever we do, there may be a transition period where we have to get used to the new version, and where we'll run into annoying problems when switching back and forth between branches. Probably best to just dive in and see how it goes... but let me know if you prefer to do that before or after the 10.0 release. :-)

File: build.xml
Comment (EreMaijala) [reply build.xml]: Of course, my mistake.

I decided to move composer.phar to downloads now while at it. :)

**Comment by EreMaijala** [comment]
@demiankatz I'd like to get this into 10.0 for proper PHP 8.3 support, but if you feel it's too risky, 10.1 would be fine too.

File: build.xml
Comment (demiankatz) [reply build.xml]: Yeah, that was a weirder mess than I had realized; worth sorting out!

**Review by demiankatz — CHANGES_REQUESTED** [review]
I've found a new problem that seems to be unique to Phing 3 -- it doesn't seem to have the same permissions as Phing 2 when deleting the cache. If I run code on the Doctrine branch in #2233, it creates a directory owned by Apache under $VUFIND_LOCAL_DIR/cache/configs to hold Doctrine-generated content (which all has 644 permissions). Running shutdown with Phing 2 clears this out with no problem, but doing it with this branch breaks the process with this message:

```
Buildfile: /usr/local/vufind/build.xml

vufind > shutdown:

   [delete] Deleting: /etc/apache2/conf-enabled/vufindtest.conf
     [exec] Result: 127
Sending stop command to Solr running on port 8983 ... waiting up to 180 seconds to allow Jetty process 19593 to stop gracefully.
     [exec] Result: 127
   [delete] Deleting: /usr/local/vufind/env.bat
   [delete] Deleting: /usr/local/vufind/env.sh
   [delete] Deleting: /usr/local/vufind/import/solrmarc.log
   [delete] Deleting directory /usr/local/vufind/solr
 [filesize] /usr/local/vufind/vufind-exception.log filesize is 0B
   [delete] Deleting: /usr/local/vufind/vufind-exception.log
   [delete] Deleting: /usr/local/vufind/vufind-mail.log
   [delete] Deleting directory /usr/local/vufind/local

BUILD FAILED
/usr/local/vufind/build.xml:1049:78 Unable to delete file /usr/local/vufind/local/cache/configs/b41555749e205d1b610f6a8eaed1702e/module-classmap-cache.php: Cannot delete /usr/local/vufind/local/cache/configs/b41555749e205d1b610f6a8eaed1702e/module-classmap-cache.php


Total time: 3.1199 seconds
```

Here's a directory listing of the offending cache:

```
dkatz@dkatz-VirtualBox:/usr/local/vufind$ ll /usr/local/vufind/local/cache/configs/b41555749e205d1b610f6a8eaed1702e/
total 352
drwxr-xr-x 2 www-data www-data   4096 Jun 18 14:39 ./
drwxr-xr-x 3 www-data www-data   4096 Jun 18 14:39 ../
-rw-r--r-- 1 www-data www-data    405 Jun 18 14:39 module-classmap-cache.php
-rw-r--r-- 1 www-data www-data 346745 Jun 18 14:39 module-config-cache.php
```

I imagine you can reproduce this problem by creating, chown-ing and chmod-ing some files without having to actually set up the Doctrine stuff. Any idea what's going on?

If we can get this sorted out, I'm not opposed to merging this close to the release date, since Phing is not normally used as part of the package unless people are pulling code from Git, so it's unlikely to break things for anyone even if there are some lingering issues that need to be addressed.

**Comment by EreMaijala** [comment]
@demiankatz Permission handling of mkdir has changed, but since the cache is not created by phing, I suspect the problem was that the extra_shutdown_cleanup task (or extra_startup_command) wasn't executed properly. If this is the case, it should be fixed now. I also changed the way e.g. `${apachectl}` is checked since boolean evaluation has changed and I'm not sure if it affects `<istrue>` as well.

**Review by demiankatz — APPROVED** [review]
Thanks, @EreMaijala!

**Pull request merged**
